src/Orleans/Runtime/AsynchQueueAgent.cs

```diff
@@ -111,6 +112,23 @@ public override void Stop()
             base.Stop();
         }
 
+        protected void DrainQueue(Action<T> action)
+        {
+            while (true)
+            {
+                T request;
+                try
+                {
+                    request = requestQueue.Take();
+                }
+                catch (InvalidOperationException)
+                {
+                    break; // Queue is empty
```

Comment:
are you using that style to detect an empty collection since using Count may be very in-efficient? 
maybe better to use TryTake instead?
catching `InvalidOperationException` just does not sounds like a robust pattern.